### PR TITLE
Add user-specific settings to database, save language preference there

### LINKED ISSUE #1381: UI should remember the user's language preference (LBR)
From LBR: "It seems that the interface doesn't remember the user's language preference. At least I've had to switch each time I've logged in."

### PR DESCRIPTION
**Pull request by okahilak**
Closes #1381

### CODE REVIEW

File: test/clj/rems/db/test_user_settings.clj [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+(use-fixtures :each rollback-db-fixture)
+
+(deftest test-user-settings
+  (users/add-user! "user1" {:eppn "user1"})
+  (testing "user settings are empty for a new user"
+    (is (= (user-settings/get-user-settings "user1")
+           {})))
+
+  (testing "add a user setting"
+    (user-settings/update-user-settings! "user1" {:language :en})
```
Comment (luontola) [inline test/clj/rems/db/test_user_settings.clj]: This tests the INSERT, but another test is needed for the UPDATE.

Comment (okahilak) [reply test/clj/rems/db/test_user_settings.clj]: Done

File: src/clj/rems/api/user_settings.clj [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  (context "/user-settings" []
+    :tags ["user-settings"]
+
+    (GET "/" []
+      :summary "Get user settings"
+      :roles #{:logged-in}
+      :return UserSettings
+      (ok (user-settings/get-user-settings (getx-user-id))))
+
+    (PUT "/update" []
```
Comment (luontola) [inline src/clj/rems/api/user_settings.clj]: This too could be "/" for a more REST style.

File: src/clj/rems/api/user_settings.clj [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      :summary "Update user settings"
+      :roles #{:logged-in}
+      :body [command UpdateUserSettingsCommand]
+      :return SuccessResponse
+      (ok (user-settings/update-user-settings! (:user-id command) command)))))
```
Comment (luontola) [inline src/clj/rems/api/user_settings.clj]: This allows anybody updating any other user's settings. The user-id parameter should probably be removed from the command, so that it'll only change the current user's settings.

File: src/clj/rems/db/user_settings.clj [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+(defn update-user-settings! [user settings]
+  (assert (and user settings) "User or user settings missing!")
+  (db/update-user-settings!
+   {:user user
+    :settings (json/generate-string {:language (:language settings)})})
+  {:success true})
+
+(defn- parse-settings [settings]
+  (let [{:keys [language]} (json/parse-string settings)]
+    {:language (keyword language)}))
```
Comment (luontola) [inline src/clj/rems/db/user_settings.clj]: Since :language is an optional key in UpdateUserSettingsCommand, how does this handle nil language? We probably need a test which sets empty settings and then reads them back.

File: src/cljs/rems/language.cljs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+(defn- update-css [language]
+  (let [localized-css (str "/css/" (name language) "/screen.css")]
+    ;; Figwheel replaces the linked stylesheet
+    ;; so we need to search dynamically
+    (doseq [element (array-seq (.getElementsByTagName js/document "link"))]
+      (when (str/includes? (.-href element) "screen.css")
+        (set! (.-href element) localized-css)))))
+
+(defn update-language [language]
+  (set! (.. js/document -documentElement -lang) language)
```
Comment (luontola) [inline src/cljs/rems/language.cljs]: This should be `(name language)`. Currently it produces `<html lang=":en">`.

File: src/cljs/rems/language.cljs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+ (fn [db _]
+   (:default-language db)))
+
+(reg-event-fx
+ ::set-language
+ (fn [{:keys [db]} [_ language]]
+   (let [user-id (get-in db [:identity :user :eppn])]
+     (if user-id
+       (rf/dispatch [:rems.user-settings/update-user-settings user-id {:language language}])
+       {:db (assoc db :language language)}))))
```
Comment (luontola) [inline src/cljs/rems/language.cljs]: `update-css` and `update-language` need to be called after updating the language here.

File: src/cljs/rems/user_settings.cljs [hunk 7]
```diff
@@ -0,0 +1,22 @@
+(ns rems.user-settings
+  (:require [re-frame.core :as rf :refer [reg-event-fx reg-event-db]]
+            [rems.language :as language]
+            [rems.status-modal :as status-modal]
+            [rems.util :refer [fetch put!]]))
+
+(reg-event-fx
+ :loaded-user-settings
+ (fn [{:keys [db]} [_ user-settings]]
+   (language/update-language (:language user-settings))
```
Comment (luontola) [inline src/cljs/rems/user_settings.cljs]: Need to check for a nil language. Currently this produces `<html lang="null">` when the user has no settings.

Comment (luontola) [reply src/cljs/rems/user_settings.cljs]: Or then make it impossible to have a nil language. It there are no user settings, then some reasonable defaults should be returned.

File: test/clj/rems/db/test_user_settings.clj [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  (:require [clojure.test :refer :all]
+            [rems.db.testing :refer [rollback-db-fixture test-db-fixture]]
+            [rems.db.user-settings :as user-settings]
+            [rems.db.users :as users]))
+
+(use-fixtures :once test-db-fixture)
+(use-fixtures :each rollback-db-fixture)
+
+(deftest test-user-settings
+  (users/add-user! "user1" {:eppn "user1"})
```
Comment (luontola) [inline test/clj/rems/db/test_user_settings.clj]: These tests would not notice it if the SQL queries were missing WHERE conditions. For that you need a second user and then check that its settings are not changed when some other user's settings are changed.

File: src/clj/rems/api/user_settings.clj (hunk 2 above)
Comment (okahilak) [reply src/clj/rems/api/user_settings.clj]: Done

File: src/clj/rems/api/user_settings.clj (hunk 3 above)
Comment (okahilak) [reply src/clj/rems/api/user_settings.clj]: Good point, done.

File: src/clj/rems/db/user_settings.clj (hunk 4 above)
Comment (okahilak) [reply src/clj/rems/db/user_settings.clj]: Done, robustified both get-user-settings and update-user-settings! and added a couple of tests.

File: src/cljs/rems/language.cljs (hunk 5 above)
Comment (okahilak) [reply src/cljs/rems/language.cljs]: Good catch, this was actually wrong in the old code as well.

File: src/cljs/rems/language.cljs (hunk 6 above)
Comment (okahilak) [reply src/cljs/rems/language.cljs]: Done.

File: test/clj/rems/db/test_user_settings.clj (hunk 8 above)
Comment (okahilak) [reply test/clj/rems/db/test_user_settings.clj]: Done.

File: src/cljs/rems/user_settings.cljs (hunk 7 above)
Comment (okahilak) [reply src/cljs/rems/user_settings.cljs]: I ended up returning default settings from get-user-settings if they are unset. It seemed like the most robust solution. Thanks for the suggestion.

**Comment by okahilak** [comment]
Thanks for the review, lots of good points.

File: resources/sql/queries.sql [hunk 9]
```diff
@@ -670,6 +670,12 @@ VALUES (:user, :userattrs::jsonb)
 ON CONFLICT (userId)
 DO UPDATE SET userAttrs = :userattrs::jsonb;
 
+-- :name update-user-settings! :!
+INSERT INTO user_settings (userId, settings)
+VALUES (:user, :settings::jsonb)
+ON CONFLICT (userId)
+DO UPDATE SET settings = :settings::jsonb;
```
Comment (luontola) [inline resources/sql/queries.sql]: If you change this to "DO NOTHING", none of the tests will detect it.

File: src/clj/rems/api/user_settings.clj [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+            [rems.db.user-settings :as user-settings]
+            [rems.util :refer [getx-user-id]]
+            [ring.util.http-response :refer :all]
+            [schema.core :as s]))
+
+(s/defschema UpdateUserSettingsCommand
+  {(s/optional-key :language) s/Keyword})
+
+(s/defschema UserSettings
+  {(s/optional-key :language) s/Keyword})
```
Comment (luontola) [inline src/clj/rems/api/user_settings.clj]: This can be made non-optional. Also the update command can be made so. Simplifies many things.

File: src/clj/rems/db/user_settings.clj [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+  (assert user "User missing!")
+  (let [settings (get-user-settings user)
+        update-language? (and language
+                              (contains? (set (env :languages)) language))]
+    (db/update-user-settings!
+     {:user user
+      :settings (json/generate-string
+                 (merge settings
+                        (when update-language?
+                          {:language language})))})
```
Comment (luontola) [inline src/clj/rems/db/user_settings.clj]: The validation of allowed languages could be extracted to its own method. Then this line would become very simple: `(merge old-settings (validate new-settings))`

Comment (okahilak) [reply src/clj/rems/db/user_settings.clj]: Done.

File: src/clj/rems/api/user_settings.clj (hunk 10 above)
Comment (okahilak) [reply src/clj/rems/api/user_settings.clj]: I was probably thinking of some other settings that are potentially changed from somewhere else and kind of preparing for partial updating. But better to have it simple for now. Done.

File: resources/sql/queries.sql (hunk 9 above)
Comment (okahilak) [reply resources/sql/queries.sql]: The test was there previously (due to your previous review comment) but got lost in the latest patch. Done.
